rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2659 @@
+
+## Using the Types in the Block Protocol
+
+Using the proposed type system for Block Protocol imposes changes on the Graph Service and how Block Schemas are defined.
+
+The Graph Service is currently specified as [this schema](https://github.com/blockprotocol/blockprotocol/blob/main/packages/%40blockprotocol/graph/src/graph-service.json).
+
+### Interfacing with properties on Entities
+
+A key change for allowing the proposed type system to work is moving away from arbitrary property keys and making use of canonical property URIs. As seen in the examples in earlier sections, properties that use simple keys will now have to point at _Property Type URIs_.
+
+**An example of an [`Entity`](https://blockprotocol.org/types/services/graph/entity) instance in the current system:**
+
+```json
+{
+  "entityId": 111,
+  "properties": {
+    "name": "Arthur Philip Dent",
+    "age": 30,
+    "planetOfOrigin": "Earth",
+    "occupation": "Intergalactic Traveler"
+  }
+}
+```
+
+**An example of an entity instance in the proposed system:**
+
+```json
+{
+  "https://blockprotocol.org/types/@blockprotocol/property-type/entity-id": 111,
+  "https://blockprotocol.org/types/@alice/property-type/name": "Arthur Philip Dent",
+  "https://blockprotocol.org/types/@alice/property-type/age": 30,
+  "https://blockprotocol.org/types/@alice/property-type/planet-of-origin": "Earth",
+  "https://blockprotocol.org/types/@alice/property-type/occupation": "Intergalactic Traveler"
+}
+```
+
+This change canonicalizes property keys, such that they uniquely identify a Property Type.
+
+**Within blocks** this impacts:
+
+- how entity fields are accessed
+- how Block Protocol messages are constructed
+- payload size of Block Protocol messages
+
+**For block authors** this impacts:
+
+- how Block Schemas are defined
+- reusability of properties (the need for some kind of type discovery)
+
+> 💡 Notice that `entityId`s are replaced by a Property Type URI that resides in the `/@blockprotocol` namespace. This is an implementation detail, and not something that is strictly dictated by the proposal.
+
+### Interfacing with Linked Entities
+
+#### Receiving Links and Linked Entities
+
+Link groups and Linked entities in the Graph Service are currently supplied outside the entity as separate objects. This behavior will stay the same, but the objects received will be of a different shape.
+
+**An example of a [`LinkGroup`](https://blockprotocol.org/types/services/graph/link-group) instance in the current system:**
+
+```json
+{
+  "sourceEntityId": "user1",
+  "path": "company",
+  "links": [
+    {
+      "sourceEntityId": "user1",
+      "destinationEntityId": "company1",
+      "path": "company"
+    }
+  ]
+}
+```
+
+**An example of a `LinkGroup` instance in the proposed system:**
+
+```json
+{
+  "sourceEntityId": 111,
```

Comment:
Is there a reason the ids have changed? If not, I'd keep them consistent so it's easier to see what the proposal actually changes

Reply:
Just an attempt to keep things simple, seems like it's proven distracting instead so we should update 👍

Reply:
Addresses here https://github.com/blockprotocol/blockprotocol/pull/352/commits/810a0fd6ca7c82655da3809bb9d5575fda297681 Thanks!